### DISCUSSION TITLE
UFCS with multiple related arguments

### DISCUSSION THREAD

**Discussion by msadeqhe** [hsutter/cppfront#811] [Suggestions]
UFCS is really expressive when we want to constrcut an object from one argument:

```cpp
person: type = {
    // ...
    operator=: (out this, name: std::string) = { /* ... */ }
    // ...
}

var1: = "Alex".person();
//  : person = "Alex";
```

Also it's nice to convert or manipulate objects with constructors:

```cpp
var2: = "Hello World!".std::string(5);
//  : = "Hello";
```

And it looks like UDL suffixes and supports multiple units too:

```cpp
var3: = 10.box().apple();
```

But when we construct an object from multiple related arguments, it significantly harms readability:

```cpp
var4: = 0.point(0, 0, true);
var5: = "Alex".person(30, country::usa);
```

Would it be possible to extend UFCS to support multiple arguments on the left side?

```cpp
var4: = (0, 0, 0).point(true);
var5: = ("Alex", 30, country::usa).person();
```

It increases readability and helpfully allows us to separate related arguments to the left side of `operator.`.

I opened two issues (#455, #463) to request this feature before, but mistakenly my suggestions contained syntax changes.

The syntax doesn't really matter, only a way to put related arguments aside is enough. Thanks.

**Comment by msadeqhe** [comment]
> But when we construct an object from multiple related arguments, it significantly harms readability:
> 
> ```c++
> var4: = 0.point(0, 0, true);
> var5: = "Alex".person(30, country::usa);
> ```
> 
> Would it be possible to extend UFCS to support multiple arguments on the left side?
> 
> ```c++
> var4: = (0, 0, 0).point(true);
> var5: = ("Alex", 30, country::usa).person();
> ```
> 
> It increases readability and helpfully allows us to separate related arguments to the left side of `operator.`.

I know it conflicts with comma operator in Cpp1, but it can work without any problem in Cpp2.

**Comment by SebastianTroy** [comment]
var4: = (0, 0, 0).point(true);
var5: = ("Alex", 30, country::usa).person();

At this point, why is this better than

var4: = point(0, 0, 0, true);
var5: = person("Alex", 30, country::usa);

I don't dislike your proposal per-se, I just want to understand why you think we can increase readability by arbitrarily splitting up arguments between before and after the function /typename

Also I understand that ufcs allows for certain templates to become more generic, does this proposal help this at all?


On 8 November 2023 06:52:10 Sadeq ***@***.***> wrote:

UFCS is really expressive when we want to constrcut an object from one argument:

person: type = {
    // ...
    operator=: (out this, name: std::string) = { /* ... */ }
    // ...
}

var1: = "Alex".person();
//  : person = "Alex";

Also it's nice to manipulate objects with constructors:

var2: = "Hello World!".std::string(5);

And it looks similar to UDL suffixes that supports multiple units:

var3: = 10.box().apple();

But when we want to construct an object from multiple arguments, it significantly harms readability:

var4: = 0.point(0, 0, true);
var5: = "Alex".person(30, country::usa);

Would it be possible to extend UFCS to support multiple arguments?

var4: = (0, 0, 0).point(true);
var5: = ("Alex", 30, country::usa).person();

It would helpfully allow us to separate related arguments to the left side of operator..

I opened two issues (#455<https://github.com/hsutter/cppfront/issues/455>, #463<https://github.com/hsutter/cppfront/issues/463>) to request this feature before, but mistakenly my suggestions contained syntax changes.

The syntax doesn't really matter, only a way to put related arguments aside is enough. Thanks.

**Reply by msadeqhe** [reply]
> I don't dislike your proposal per-se, I just want to understand why you think we can increase readability by arbitrarily splitting up arguments between before and after the function /typename

According to [P3021r0](https://open-std.org/jtc1/sc22/wg21/docs/papers/2023/p3021r0.pdf) paper from @hsutter, UFCS has the following benefits:

3.1.1. It’s friendlier to editors and tools
3.1.2. It’s friendlier to programmers: Discoverability
3.1.3. It more directly corresponds to what programs do: State transition
3.1.4. The existing-practice workarounds already simulate extending the member call syntax, not the nonmember syntax

If I want to construct a 3D point, and `|` is the cursor in the following example:

```cpp
var1: = (0, 0, 0).|
```

The editor will show me the appropriate list of functions and types (3.1.1). The editor will find types and functions which can be called with those parameters, and it will show me a list of them. In this example, it may show me a list with `point`, `matrix`, etc.

Also as Herb wrote in that paper (3.1.2):

> This illustrates why the member call syntax is generally friendlier to programmers. I believe the “start with the
object” point of view is a much more usable programming model because it directly enables answering the con-
stant question of how to discover “what can I do next?”. What you can do next (functions) depends on what you
have now to do it with (your objects and variables).

In this example, our object has 3 parts `(0, 0, 0)` instead of 1 part `"Alex"`.

And another important case is _state transition_ (3.1.3), I as a programmer prefer to write the object and then I do something to that. In this example, our object has 3 parts `(0, 0, 0)`.

And for the last case (3.1.4):

```cpp
(first(), second()).third();

third(first(), second());
```

Infact, my suggestion is motiviated from Herb's [P3021r0](https://open-std.org/jtc1/sc22/wg21/docs/papers/2023/p3021r0.pdf) paper that makes me think why we still need UFCS with multiple related arguments.

> Also I understand that ufcs allows for certain templates to become more generic, does this proposal help this at all?

I mean, the only thing that matters in my suggestion is readability and toolability, and it works in generic programming too (because of UFCS):

```cpp
fnc1: (a: i32, b: i32, c: bool) = { /* ... */ }

(0, 0).fnc1(true);
```

So `(0, 0).fnc1(true)` calls non-member functions too. The separated arguments would make it clear that what was the intention of the programmer in generic programming. So yes, it helps in generic programming too. Thanks.

**Reply by msadeqhe** [reply]
I think generally member function call is preferred over non-member function call, because member function call is UFCS in Cpp2, so UFCS can be used for every function call including functions/constructors with multiple related arguments (such as `(0, 0, 0)`).

If I use non-member function call syntax just in case I want to construct an object with multiple related arguments, I will lose the benefits of UFCS. For example if I use:

```cpp
var4: = point(0, 0, 0, true);
var5: = person("Alex", 30, country::usa);
```

I will lose the benefits of UFCS in generic programming in addition to decreasing readability and toolability.

**Comment by SebastianTroy** [comment]
Oh I'm already well in favour of UFCS, and I agree that your multiple arguments allows for nicer discovery of types pertaining to the variables, but what programmer intent is portrayed when viewing

(a, b).f(c);

Vs

a.f(b, c);

Autocomplete is already the work of a tool, so extending it to work with "(a, b, c)." doesn't necessarily require a language feature, as it could easily auto complete to "type(a, b, c)"

Basically my point is, do we need to visually complicate the language further, when the feature you suggest is for a tool rather than the language?



On 8 November 2023 10:21:10 Sadeq ***@***.***> wrote:

I don't dislike your proposal per-se, I just want to understand why you think we can increase readability by arbitrarily splitting up arguments between before and after the function /typename

According to P3021r0 paper from @hsutter<https://github.com/hsutter>, UFCS has the following benefits:

3.1.1. It’s friendlier to editors and tools
3.1.2. It’s friendlier to programmers: Discoverability
3.1.3. It more directly corresponds to what programs do: State transition
3.1.4. The existing-practice workarounds already simulate extending the member call syntax, not the nonmember syntax

If I want to construct a 3D point, and | is the cursor in the following example:

var1: = (0, 0, 0).|

The editor will show me the appropriate list of functions and types (3.1.1). The editor will find types and functions which can be called with those parameters, and it will show me a list of them. In this example, it may show me a list with point, matrix, etc.

Also as Herb wrote in that paper (3.1.2):

This illustrates why the member call syntax is generally friendlier to programmers. I believe the “start with the
object” point of view is a much more usable programming model because it directly enables answering the con-
stant question of how to discover “what can I do next?”. What you can do next (functions) depends on what you
have now to do it with (your objects and variables).

In this example, our object has 3 parts (0, 0, 0) instead of 1 part "Alex".

And the most important case in my example is state transition (3.1.3), I as a programmer prefer to write the object and then I do something to that.

And for the last case (3.1.4):

(first(), second()).third();

third(first(), second());

Infact, my suggestion is motiviated from Herb's P3021r0 paper that makes me think why we still need UFCS with multiple related arguments.

Also I understand that ufcs allows for certain templates to become more generic, does this proposal help this at all?

I mean, the only thing that matters in my suggestion is readability and toolability, and it works in generic programming too (because of UFCS):

fnc1: (a: i32, b: i32, c: bool) = { /* ... */ }

(0, 0).fnc1(true);

So (0, 0).fnc1(true) calls non-member functions too. The separated arguments would make it clear that what was the indention of the programmer in generic programming. So yes, it helps in generic programming too. Thanks.

**Reply by msadeqhe** [reply]
You're right. My suggestion is visually useful for constructors only. I think more about it, if I don't find a useful feature from this syntax, I will close this discussion. Thanks.

**Reply by msadeqhe** [reply]
> Autocomplete is already the work of a tool, so extending it to work with "(a, b, c)." doesn't necessarily require a language feature, as it could easily auto complete to "type(a, b, c)"

Replace `type` with `function` in `function(a, b, c)` and we lose UFCS in generic programming, so it will never call member functions. Therefore a language feature is required to have `(a, b, c).function()`.

In addition to readability and toolability, I'm thinking about use cases and why `(a, b, c).function()` would help us to write short/nice code.

**Comment by JohelEGP** [comment]
I think this might encourage writing inside-out code again.
Consider slide 127 of <https://github.com/codereport/Content/blob/main/Talks/2023-07-CppNorth/New%20Algorithms%20in%20C%2B%2B23/New%20Algorithms%20in%20C%2B%2B23.pdf>.

```C++
auto sushi_for_two(std::vector<int> sushi) { // clang-format off
  return sushi
      |> zip_with(_neq_, $, $ | drop(1))
      |> zip($, iota(1))
      |> filter($, _fst)
      |> values($)
      |> concat(single(0), $, single(sushi.size()))
      |> zip_with(_sub_, $ | drop(1), $)
      |> zip_with(_min_, $, $ | drop(1))
      |> std::ranges::max($) * 2;
} // clang-format on
```

Line `|> concat(single(0), $, single(sushi.size()))` has the input object as the second argument.
That's `.call(:(x) concat(single(0), x, sushi.size().single()))`
with the `call` from <https://github.com/hsutter/cppfront/issues/741#issuecomment-1754194152>.
With multiple-argument UFCS, it would be possible to break the linearity of the code from
```Cpp2
sushi_for_two: (sushi: std::span<int>) //
  sushi                                //
    .call(:(x) zip_with(_neq_, x, x.drop(1)))
    .zip(iota(1))
    .filter(_fst)
    .values()
    .call(:(x) concat(single(0), x, sushi.size().single()))
    .call(:(x) zip_with(_sub_, x.drop(1), x))
    .call(:(x) zip_with(_min_, x, x.drop(1)))
    .std::ranges::max() *
  2;
```
to
```Cpp2
sushi_for_two: (sushi: std::span<int>) //
  (single(0), sushi                    //
                .call(:(x) zip_with(_neq_, x, x.drop(1)))
                .zip(iota(1))
                .filter(_fst)
                .values())
    .concat(sushi.size().single())
    .call(:(x) zip_with(_sub_, x.drop(1), x))
    .call(:(x) zip_with(_min_, x, x.drop(1)))
    .std::ranges::max() *
  2;
```

**Reply by msadeqhe** [reply]
Thanks. If we could declare variable within parentheses, it would be like this:

```cpp
sushi_for_two: (sushi: std::span<int>) (
    _min_,
    a: = (
        _sub_,
        (b: = (
            single(0),
            (_neq_, c: = sushi, c.drop(1))
             .zip_with()
             .zip(iota(1))
             .filter(_fst)
             .values(),
            sushi
             .size()
             .single()
        ).concat()).drop(1),
        b
    ).zip_with(),
    a.drop(1)
).zip_with()
 .std::ranges::max() * 2;
```

***EDIT:*** But there is a difference, in my example it doesn't use lambdas. It directly calls the functions.

**Reply by msadeqhe** [reply]
If Cpp2 had keyword `value`, `prev` or `it` to refer to the value of the left operand of `operator.`, the example would be changed to:

```cpp
sushi_for_two: (sushi: std::span<int>)
  sushi
    .zip_with(_neq_, it, it.drop(1))
    .zip(iota(1))
    .filter(_fst)
    .values()
    .concat(single(0), it, sushi.size().single())
    .zip_with(_sub_, it.drop(1), it)
    .zip_with(_min_, it, it.drop(1))
    .std::ranges::max() * 2;
```

With keyword `value`, `prev` or `it`, we don't have to use lambdas, we can directly use UFCS.

Also it would be helpful in every expression too. `it` means the left operand in everywhere:

```cpp
a: = 10;

var1: = (a * a + 2) * it / 2;

// The above line is equal to the following lines:
xx: = a * a + 2;
var2: = xx * xx / 2;
```

***EDIT:*** But unfortunately `it` doesn't work with nested `operator.` in UFCS:

```cpp
x: = a.b(c, m.d(it));
```

`it` in `m.d(it)` doesn't refer to `a`, instead it refers to `m`.

***EDIT:*** I've removed all the parts which was about reflections and solving the nested `operator.` problem. It wasn't a correct solution. sry.

**Comment by msadeqhe** [comment]
Briefly, I suggest to make ...

```cpp
(obj, a1, a2, ...).call(b1, b2, ...)
```

... a syntax sugar to ...

```cpp
obj.call(a1, a2, ..., b1, b2, ...)
```

Therefore if `call` is a type, it would call the constructor without `this` argument, otherwise it would call a non-member function:

```cpp
TYPE(/* out this */, obj, a1, a2, ..., b1, b2, ...)
FUNCTION(obj, a1, a2, ..., b1, b2, ...)
```

But if `call` is a member function of `obj`, it would call it with `obj` as `this` argument, otherwise it would call a non-member funcion:

```cpp
obj.FUNCTION(a1, a2, ..., b1, b2, ...)
FUNCTION(obj, a1, a2, ..., b1, b2, ...)
```

So the behavior of it depends on if `call` is a type or a member funcion.

As you see, `(obj, a1, a2, ...).call(b1, b2, ...)` is more readable when `call` is a type, because `obj` is similar to `a1, a2, ...` arguments. But when `call` is a function, it may be misleading, because `obj` is alongside `a1, a2, ...` arguments while they are different.

So if we use UFCS, we can argue that `(a, b, c).call()` is the right way to call constructors to create objects, and `a.call(b, c)` is the right way to call functions. In generic programming, both of these syntaxes are helpful. Because I as a programmer want to create an object, and if my code works as intended with a member function or non-member function too, also I would be happy.

On the other hand, if for readability reasons I change `x.point(y, z)` to `point(x, y, z)`, I lose the power of UFCS in generic programming, because it won't call member functions if `point` isn't a type. But if I change it to `(x, y, z).point()`, I have the power of UFCS in generic programming, because it will call member function `point` for me if `x` has `point` member function.

As a result, `(x, y, z).something(...)` is necessary if we want to use UFCS in generic programming while the code loses its readability with `x.something(y, z, ...)`.

**Comment by AbhinavK00** [comment]
Here's why I think this is not worth it. UFCS unified two notations that ALREADY EXISTED but here you are making a new one. This does not help in writing code in any way, because there's no code out there that has this new syntax in it. Also, the advantages that Herb mentions in his paper do not apply here. This is not friendlier to editor, tools or programmer. Most of that friendliness came because dot syntac already existed but this one doesn't.
You gotta understand that UFCS was about extending an already existing syntax but this one is about making a new syntax which makes all the difference. 
_plus it requires me to type TWO sets of parens_

**Reply by msadeqhe** [reply]
Fair point. So if I want to create a 3D point and if `x.point(y, z, true)` is not readable, I have to write it like ...
`point(x, y, z, true)`
..., but this won't call _member_ functions in generic code. Therefore the solution (alternatively) would be to allow ...
`point(x, y, z, true)`
... to call _member_ functions as described in part 3.2.3 from [P3021r0](https://open-std.org/jtc1/sc22/wg21/docs/papers/2023/p3021r0.pdf).

***EDIT:*** I've corrected _non-member_ to _member_.

**Reply by AbhinavK00** [reply]
Not sure, but I don't think UFCS works with contructors. If I'm correct, then the only way to currently contruct a point will be
```
Point(x,y,z,true);
```

**Reply by JohelEGP** [reply]
This works (<https://cpp2.godbolt.org/z/8cWvfjKG9>): `main: () = { _ = "OK".std::string(); }`.
